### PR TITLE
Fix metadata caching with Hive ACID

### PR DESCRIPTION
**Pull request by dain**
## Description

Fix shared metadata caching with Hive ACID tables.

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Hive
* Fix shared metadata caching with Hive ACID tables. ({issue}`11443`)
```

### CODE REVIEW

**Comment by dain** [comment]
This still needs more work

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/AcidTransactionOwner.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.metastore;
+
+import static java.util.Objects.requireNonNull;
+
+// this is just a type safe wrapper to make the APIs cleaner
```
Comment (jirassimok) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/AcidTransactionOwner.java:18]: With this in mind, I would definitely add the ID to this class, and call it `TransactionInfo` or something like that.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 2]
```diff
@@ -1322,7 +1323,8 @@ private HiveTransaction makeHiveTransaction(ConnectorSession session, Function<L
         long heartbeatInterval = configuredTransactionHeartbeatInterval
                 .map(Duration::toMillis)
                 .orElseGet(this::getServerExpectedHeartbeatIntervalMillis);
-        long transactionId = delegate.openTransaction();
+        // TODO consider adding query id to the owner
+        long transactionId = delegate.openTransaction(new AcidTransactionOwner(session.getIdentity().getUser()));
```
Comment (jirassimok) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java:1327]: `session.getUser()`

**Comment by losipiuk** [comment]
RN should be expressed in end-user terms of end-user visible effects.

**Review by jirassimok — COMMENTED** [review]
This is a fairly straightforward change that seems to make sense, though I can't really judge if it's the best way to do it.

Particularly, in its current form, `AcidTransactionOwner` is a near-duplicate of `HiveIdentity`, which was removed from `HiveMetastore` not long ago.

**Comment by dain** [comment]
I have some more refactorings in another branch.  This PR is focused on fixing the broken caching behavior.

**Comment by losipiuk** [comment]
What would happen if we did not set the user `LockRequestBuilder` at all.
I see that Hive code does:
```
        if (user == null) {
            user = "unknown";
        }
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java [hunk 3]
```diff
@@ -1574,15 +1575,15 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
     }
 
     @Override
-    public long openTransaction(HiveIdentity identity)
+    public long openTransaction(HiveIdentity identity, AcidTransactionOwner transactionOwner)
```
[the remark this answers is not in the document]
Comment (dain) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:1578]: It is for now.... I have changes in another branch to clean this stuff up, but wanted to keep this simple

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/AcidTransactionOwner.java (hunk 1 above)
Comment (dain) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/AcidTransactionOwner.java:18]: I have this another branch for later :)

**Review by losipiuk — APPROVED** [review]
LGTM as a minimal change to be sure user identity is properly propagated from session to ACID related metastore methods.

**Comment by dain** [comment]
> What is the nature of breakage?

You get an IllegalArgumentException when starting a read or write ACID transaction, so caching does not work at all with ACID.

> What would happen if we did not set the user `LockRequestBuilder` at all? I see that Hive code does:
> 
> ```
>         if (user == null) {
>             user = "unknown";
>         }
> ```

I wish I knew... I had to code all of this blind.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 4]
```diff
@@ -1322,7 +1323,8 @@ private HiveTransaction makeHiveTransaction(ConnectorSession session, Function<L
         long heartbeatInterval = configuredTransactionHeartbeatInterval
                 .map(Duration::toMillis)
                 .orElseGet(this::getServerExpectedHeartbeatIntervalMillis);
-        long transactionId = delegate.openTransaction();
+        // TODO consider adding query id to the owner
```
[the remark this answers is not in the document]
Comment (dain) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java:1326]: Not sure what you mean?  We could add the query id to the passed in owner String so you know which query owns the acid transaction

**Review by findepi — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java [hunk 5]
```diff
@@ -65,6 +66,7 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
 
         long lockId = thriftMetastore.acquireTableExclusiveLock(
                 identity,
+                new AcidTransactionOwner(session.getUser()),
```
[the remark this answers is not in the document]
Comment (dain) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java:69]: We can fix this later, but I don't see how this is not ACID.  Weren't the APIs added for ACID exclusively?

[the remark this answers is not in the document]
Comment (dain) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java:69]: Honestly, I think we should make these strings be `queryid-username` or just `queryid`

**Comment by dain** [comment]
The test failures appear to be flakes, so merging.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java:69]: > We can fix this later, but I don't see how this is not ACID. Weren't the APIs added for ACID exclusively?

`acquireTableExclusiveLock` method name doesn't say "ACID" and here it's used for non-ACID purposes

sure, we can fix it later -- mind filing an issue?

**Pull request merged**